kernel/kernel-api/src/main/java/io/delta/kernel/internal/actions/Protocol.java

```diff
@@ -221,6 +223,25 @@ public Set<TableFeature> getImplicitlyAndExplicitlySupportedFeatures() {
     return supportedFeatures;
   }
 
+  /**
+   * Get the set of reader writer features that are both implicitly and explicitly supported by the
+   * protocol. Usually, the protocol has either implicit or explicit features, but not both. This
+   * API provides a way to get all enabled reader writer features. It doesn't return any writer only
+   * features.
+   */
+  public Set<TableFeature> getImplicitlyAndExplicitlySupportedReaderWriterFeatures() {
+    return Stream.concat(
+            // implicit supported features
+            TABLE_FEATURES.stream()
+                .filter(
+                    f ->
+                        supportsReaderFeatures
+                            && f.minReaderVersion() <= this.getMinReaderVersion()),
```

Comment:
I'm confused by this... doesn't this return nothing for all legacy protocol versions since `supportsReaderFeatures=false`?

Reply:
Also, maybe this can just be

```
if (supportsReaderFeatures) {
    // return explicit features
} else {
   // return implicit features
}
```

Reply:
yeah shouldn't this be `!supportsReaderFeatures && ...`? that's what I see in `getImplicitlySupportedFeatures`

Reply:
also wondering why this isn't caught in tests? maybe I'm missing something obvious not sure sorry I'm confused :/